### PR TITLE
Added fs2.async.SyncRef

### PR DESCRIPTION
**Pull request by mpilquist**
...a lightweight alternative to Ref backed by an AtomicReference.

### CODE REVIEW

File: core/shared/src/main/scala/fs2/async/async.scala [hunk 1]
```diff
@@ -94,7 +96,7 @@ package object async {
    * bound. The inner `F[A]` will block until the result is available.
    */
   def start[F[_], A](f: F[A])(implicit F: Effect[F], ec: ExecutionContext): F[F[A]] =
-    ref[F, A].flatMap { ref => ref.setAsync(F.shift(ec) *> f).as(ref.get) }
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/async/async.scala]: This shift was redundant with the one in `setAsync`

File: core/shared/src/main/scala/fs2/async/RefOps.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package fs2.async
+
+/**
+ * Operations on concurrent, mutable reference types.
+ */
+trait RefOps[F[_], A] {
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/async/RefOps.scala]: Note there's no `setAsync` operation, as that's not possible to implement on `SyncRef`

File: core/shared/src/main/scala/fs2/async/Ref.scala [hunk 3]
```diff
@@ -204,7 +186,7 @@ object Ref {
 
   /** Creates an asynchronous, concurrent mutable reference, initialized to `a`. */
   def initialized[F[_]: Effect, A](a: A)(implicit ec: ExecutionContext): F[Ref[F,A]] =
-    uninitialized[F, A].flatMap(r => r.setAsyncPure(a).as(r))
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/async/Ref.scala]: This was a bug or performance degradation depending on perspective

**Comment by SystemFw** [comment]
It mostly looks good to me. I have one question regarding the semantics though: as it stands, `Ref` and `SyncRef` have different semantics wrt to errors. I.e. on a Ref, `set(F.fail(ex))` will succeed, and the subsequent `get` will fail, whereas `set(F.fail(ex))` will fail on a SyncRef.
In other words, a SyncRef is a mutable reference for As, whereas a Ref is a mutable reference for `Throwable or A` (which puzzled me at first).
Is this intended/desirable?

**Comment by SystemFw** [comment]
I'd go with option 1.

(I guess my preference would be to have Ref to also not have that behaviour, but that's a deeper change that's best left out of this PR)

**Comment by mpilquist** [comment]
Hm, what about leaving `setSync` on `RefOps` but changing its behavior such that a failure in `fa` fails result instead of setting a failure in to the ref? Then the only way to cause `get` to throw would be a failure in `fa` passed to `setAsync`, which is understandable since the task is forked before `fa` is run.

**Comment by mpilquist** [comment]
OK, fixed in https://github.com/functional-streams-for-scala/fs2/pull/986/commits/f05db8818eb85bace11f6c9554034140c95df618

**Comment by SystemFw** [comment]
Well, I thought the whole point of having `setSync` as a separate operation is that it behaves differently than flatMap then set.

(Otoh, I have tried not too hard once to remove the Throwable from Ref altogether)

**Comment by mpilquist** [comment]
Alright, that's convincing. Just pushed a fix.

**Comment by SystemFw** [comment]
:+1: for merging this

If you want to have a look at how Ref would look like with no throwable (in general, not for the sake of this PR), I have a (very probably buggy) version [here](https://github.com/SystemFw/fs2/blob/exp/failed/ref-no-throwable/core/shared/src/main/scala/fs2/async/Ref.scala)

I tacked this while trying to add `modifyF` (which I think can be done with a new message in Actor, or modifying the TrySet message)

**Pull request merged**